src/watchdog.js

```diff
@@ -0,0 +1,393 @@
+		*
+		* @private
+		* @member {HTMLElement|String} #_elementOrData
+		*/
+
+		/**
+		* The editor configuration.
+		*
+		* @private
+		* @member {Object|undefined} #_config
+		*/
+	}
+
+	/**
+	 * The current editor instance.
+	 *
+	 * @type {module:core/editor/editor~Editor}
+	 */
+	get editor() {
+		return this._editor;
+	}
+
+	/**
+	 * Sets the function that is responsible for editor creation.
+	 * It accepts functions that returns promises.
+	 *
+	 * 		watchdog.setCreator( ( el, config ) => ClassicEditor.create( el, config ) );
+	 *
+	 * @param {Function} creator
+	 */
+	setCreator( creator ) {
+		this._creator = creator;
+	}
+
+	/**
+	 * Sets the function that is responsible for editor destruction.
+	 * It accepts function that returns a promise or undefined.
+	 *
+	 *		watchdog.setDestructor( editor => editor.destroy() );
+	 *
+	 * @param {Function} destructor
+	 */
+	setDestructor( destructor ) {
+		this._destructor = destructor;
+	}
+
+	/**
+	 * Restarts the editor instance. This method is also called whenever an editor error occurs.
+	 * It fires the `restart` event.
+	 *
+	 * @fires restart
+	 * @returns {Promise.<module:core/editor/editor~Editor>}
+	 */
+	restart() {
+		this._throttledSave.flush();
+
+		return Promise.resolve()
+			.then( () => this.destroy() )
+			.then( () => {
+				if ( typeof this._elementOrData === 'string' ) {
+					return this.create( this._data, this._config );
+				}
+
+				const updatedConfig = Object.assign( {}, this._config, {
+					initialData: this._data
+				} );
+
+				return this.create( this._elementOrData, updatedConfig );
+			} )
+			.then( () => {
+				this.fire( 'restart' );
+			} );
+	}
+
+	/**
+	 * Creates a watched editor instance using the creator passed to {@link #setCreator} method.
+	 *
+	 * @param {HTMLElement|String} elementOrData
+	 * @param {module:core/editor/editorconfig~EditorConfig} [config]
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	create( elementOrData, config ) {
+		if ( !this._creator ) {
+			/**
+			 * @error watchdog-creator-not-defined
+			 *
+			 * The watchdog creator is not defined, define it using `watchdog.setCreator()`.
+			 */
+			throw new CKEditorError(
+				'watchdog-creator-not-defined: The watchdog creator is not defined, define it using `watchdog.setCreator()`.',
+				{}
+			);
+		}
+
+		if ( !this._destructor ) {
+			/**
+			 * @error watchdog-destructor-not-defined
+			 *
+			 * The watchdog destructor is not defined, define it using `watchdog.setDestructor()`.
+			 */
+			throw new CKEditorError(
+				'watchdog-destructor-not-defined: The watchdog destructor is not defined, define it using `watchdog.setDestructor()`',
+				{}
+			);
+		}
+
+		this._elementOrData = elementOrData;
+		this._config = config;
+
+		return Promise.resolve()
+			.then( () => this._creator( elementOrData, config ) )
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorHandler );
```

Comment:
We're only handling `CKEditorError`s here. All other crashes will not be caught. This limitation should be explained in the Watchdog documentation.